library/common/extensions/filters/http/assertion/filter.h

```diff
@@ -0,0 +1,50 @@
+#pragma once
+
+#include "envoy/http/filter.h"
+
+#include "extensions/common/tap/tap_matcher.h"
+#include "extensions/filters/http/common/pass_through_filter.h"
+
+#include "library/common/extensions/filters/http/assertion/filter.pb.h"
+
+namespace Envoy {
+namespace Extensions {
+namespace HttpFilters {
+namespace Assertion {
+
+class AssertionFilterConfig {
+public:
+  AssertionFilterConfig(
+      const envoymobile::extensions::filters::http::assertion::Assertion& proto_config);
+
+  Extensions::Common::Tap::Matcher& rootMatcher() const;
+  size_t matchersSize() const { return matchers_.size(); }
+
+private:
+  std::vector<Extensions::Common::Tap::MatcherPtr> matchers_;
+};
+
+typedef std::shared_ptr<AssertionFilterConfig> AssertionFilterConfigSharedPtr;
+
+/**
+ * Filter to assert expectations on HTTP requests.
+ */
+class AssertionFilter final : public Http::PassThroughFilter {
+public:
+  AssertionFilter(AssertionFilterConfigSharedPtr config);
+
+  // StreamDecoderFilter
+  Http::FilterHeadersStatus decodeHeaders(Http::RequestHeaderMap& headers,
+                                          bool end_stream) override;
+  Http::FilterDataStatus decodeData(Buffer::Instance& data, bool end_stream) override;
+  Http::FilterTrailersStatus decodeTrailers(Http::RequestTrailerMap& trailers) override;
+
```

Comment:
Can we validate response as well? Do matchers work on responses?

The current implementation generates 200s on end_stream, but that could easily be handled by a different upstream filter. Meanwhile it would allow testing of, e.g., modification of responses via platform filters.

Reply:
Yeah the matcher can match responses. My thought was that for the initial commit I would do request validation as that is enough expressiveness to fulfill the e2e tests I had in mind for client->bridge->core interactions. The filter validates the request path and the client can validate what is returned on the response path.

So my plan was to land this, and then as you start writing tests we can modify the filter for added complexity.

Reply:
Sounds good. 👍